File: boson-core/src/main/scala/io/zink/boson/bson/bsonImpl/BosonInjector.scala
package io.zink.boson.bson.bsonImpl

import java.nio.ByteBuffer
import java.util.concurrent.CompletableFuture
import java.util.function.Function

import io.zink.boson.bson.Boson
import io.zink.boson.bson.bsonPath.{Interpreter, Program, TinyLanguage}
import io.zink.boson.bson.bsonValue.{BsBoson, BsException, BsObject, BsValue}

import scala.compat.java8.FunctionConverters._


class BosonInjector[T](expression: String, injectFunction: Function[T, T]) extends Boson {

  val anon: T => T = injectFunction.asScala

  def parseInj[K](netty: BosonImpl, injectFunction: K => K , expression: String):BsValue = {
    val parser = new TinyLanguage
    try{
      parser.parseAll(parser.program, expression) match {
        case parser.Success(r,_) =>
          new Interpreter(netty, r.asInstanceOf[Program], Option(injectFunction)).run()
        case parser.Error(msg, _) =>
          BsObject.toBson(msg)
        case parser.Failure(msg, _) =>
          BsObject.toBson(msg)
      }
    }catch {
      case e: RuntimeException => BsObject.toBson(e.getMessage)
    }
  }

  override def go(bsonByteEncoding: Array[Byte]): CompletableFuture[Array[Byte]] = {
    val boson:BosonImpl = new BosonImpl(byteArray = Option(bsonByteEncoding))
    val future: CompletableFuture[Array[Byte]] =
      CompletableFuture.supplyAsync(() =>{
      val r: Array[Byte] = parseInj(boson, anon, expression) match {
        case ex: BsException => println(ex.getValue)
          bsonByteEncoding
        case nb: BsBoson =>
          nb.getValue.getByteBuf.array()
        case _ =>
          bsonByteEncoding
      }
      r
    })
    future
  }

  override def go(bsonByteBufferEncoding: ByteBuffer): CompletableFuture[ByteBuffer] = {
    val boson: BosonImpl = new BosonImpl(javaByteBuf = Option(bsonByteBufferEncoding))
    val future: CompletableFuture[ByteBuffer] =
    CompletableFuture.supplyAsync(() =>{
      val r:ByteBuffer = parseInj(boson, anon, expression) match {
        case ex: BsException => println(ex.getValue)
          bsonByteBufferEncoding
        case nb: BsBoson => nb.getValue.getByteBuf.nioBuffer()
        case _ =>
          bsonByteBufferEncoding
      }
      r
    })
    future
  }

  override def fuse(boson: Boson): Boson = new BosonFuse(this,boson)
}

File: boson-java/src/main/java/io/zink/boson/BosonInjector.java
package io.zink.boson;

import io.zink.boson.bson.bsonImpl.BosonImpl;
import io.zink.boson.bson.bsonPath.Interpreter;
import io.zink.boson.bson.bsonPath.Program;
import io.zink.boson.bson.bsonPath.TinyLanguage;
import io.zink.boson.bson.bsonValue.*;
import io.zink.bosonInterface.Boson;
import scala.Option;
import scala.util.parsing.json.Parser;
import scala.compat.java8.FunctionConverters.*;
import java.nio.ByteBuffer;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;
import static scala.Option.*;

public class BosonInjector<T> implements Boson {
    private String expression;
    private Function<T, T> injectFunction;

    public BosonInjector(String expression, Function<T, T> injectFunction) {
        this.expression = expression;
        this.injectFunction = injectFunction;
    }
    private Function<String,BsValue> writer = BsException$.MODULE$::apply;

    private BsValue parseInj(BosonImpl netty, Function<T,T> injectFunc, String expression){
        TinyLanguage parser = new TinyLanguage();
        try{
            Parser.ParseResult pr = parser.parseAll(parser.program(), expression);
            if(pr.successful()){
                Interpreter interpreter = new Interpreter(netty, (Program) pr.get(), apply(injectFunc));
                BsValue res = interpreter.run();

                return res;
            }else{
                return BsObject$.MODULE$.toBson("Error inside interpreter.run() ", Writes$.MODULE$.apply1(writer));
            }
        }catch (RuntimeException e){
            return BsObject$.MODULE$.toBson(e.getMessage(), Writes$.MODULE$.apply1(writer));
        }
    };

    @Override
    public CompletableFuture<byte[]> go(byte[] bsonByteEncoding) {
        Option opt = apply(bsonByteEncoding);
        Option e = empty();
        BosonImpl boson = new BosonImpl(opt, e, e);
        CompletableFuture<byte[]> future =
                CompletableFuture.supplyAsync(() -> {
                    BsValue res =  parseInj(boson, injectFunction, expression);
                    try{
                        switch (res.getClass().getSimpleName()){
                            case "BsException": return bsonByteEncoding;
                            case "BsBoson": return ((BsBoson) res).getValue().getByteBuf().array();
                            default:  return bsonByteEncoding;
                        }


                    }catch(ClassCastException ex){
                        System.out.println(ex.getMessage());
                        return bsonByteEncoding;
                    }
                });
        return future;
    }

    @Override
    public CompletableFuture<ByteBuffer> go(ByteBuffer bsonByteBufferEncoding) {
        Option opt = apply(bsonByteBufferEncoding);
        Option e = empty();
        BosonImpl boson = new BosonImpl(opt, e, e);
        CompletableFuture<ByteBuffer> future =
                CompletableFuture.supplyAsync(() -> {
                    BsValue res =  parseInj(boson, injectFunction, expression);
                    switch (res.getClass().getSimpleName()){
                        case "BsException": return bsonByteBufferEncoding;
                        case "BsBoson": return ((BsBoson) res).getValue().getByteBuf().nioBuffer();
                        default:  return bsonByteBufferEncoding;
                    }
                });
        return future;
    }

    @Override
    public Boson fuse(Boson boson) {
        return new BosonFuse(this,boson);
    }
}
